### PR TITLE
PLM DTI example support for large curated DTI data

### PR DESCRIPTION
**Pull request by alex-golts**

### CODE REVIEW

File: fusedrug/utils/samplers/subset_sampler.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        elif isinstance(sample_ids, list):
+            self.indices = sample_ids
+
+        self.shuffle = shuffle
+        self.epoch = 0
+
+    def __iter__(self) -> None:
+        # offset for the current epoch
+        offset = self.epoch * self.num_samples_per_epoch
+        if offset + self.num_samples_per_epoch >= self.dataset_len:
```
Comment (mosheraboh) [inline fusedrug/utils/samplers/subset_sampler.py:53]: So you are dropping the last part of the data?
Why not do cyclic implementation?

File: fusedrug/utils/samplers/subset_sampler.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import torch
+import numpy as np
+from fuse.data.datasets.dataset_base import DatasetBase
+from typing import Optional, Union
+
+
+class SubsetSampler(torch.utils.data.Sampler):
```
Comment (mosheraboh) [inline fusedrug/utils/samplers/subset_sampler.py:7]: Is this similar to sub_epoch_sampler we have in bmfmcore?

File: fusedrug/utils/samplers/balanced_df_sampler.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            else:
+                print(f"found {len(indices)} cases of class {curr_class}")
+            self.indices_per_class.append(indices)
+            self.pointers.append(indices.shape[0])
+
+            minibatches_needed_to_see_all.append(
+                int(np.ceil(indices.shape[0] / curr_count))
+            )
+
+        smallest_class_index = np.argmin([len(x) for x in self.indices_per_class])
```
Comment (mosheraboh) [inline fusedrug/utils/samplers/balanced_df_sampler.py:56]: Shouldn't you look for the max in  minibatches_needed_to_see_all?

File: fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/losses.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import torch
+import torch.nn.functional as F
+
+## TODO: move to fuse core
+
+
+def focal_loss(
```
Comment (mosheraboh) [inline fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/losses.py:7]: is it the same focal_loss as in bmfm?

File: fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/test.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    # Load DataLoaders (wrapped by FuseMedML)
+    _, _, test_dataloader, cfg = data.get_dataloaders(cfg, test_mode=True)
+
+    model = plm_dti.PLM_DTI_Module(cfg)
+
+    trainer = pl.Trainer(
+        default_root_dir=cfg.experiment.dir,
+        devices=1,
+        accelerator="gpu",
+        benchmark=True,
```
Comment (mosheraboh) [inline fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/test.py:47]: What benchmark does?

**Review by mosheraboh — APPROVED** [review]
Looks good!
Minors inline.

File: fusedrug/utils/samplers/subset_sampler.py (hunk 1 above)
Comment (alex-golts) [reply fusedrug/utils/samplers/subset_sampler.py:53]: yes, it was under the assumption that the data is sufficiently large and it's not critical (and in some cases even not all data will be seen anyway)

File: fusedrug/utils/samplers/subset_sampler.py (hunk 2 above)
Comment (alex-golts) [reply fusedrug/utils/samplers/subset_sampler.py:7]: I think that it's a similar concept. this was done before I was aware of the bmfm one/perhaps before it existed

File: fusedrug/utils/samplers/balanced_df_sampler.py (hunk 3 above)
Comment (alex-golts) [reply fusedrug/utils/samplers/balanced_df_sampler.py:56]: every batch will contain elements from all classes, so the total number of minibatches is defined according to the rarest class. and then when there are no more examples in the rarest class, the "epoch" ends, and optionally reshuffles.
you could think of different logics, but I think this also makes sense. and there's a test for this sampler that verified the expected behavior

File: fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/losses.py (hunk 4 above)
Comment (alex-golts) [reply fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/losses.py:7]: it's a different implementation. this one supports both logit and probability inputs, and one hot labels. I also had some problems with the one in bmfm but didn't do a full comparison to understand the root cause

File: fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/test.py (hunk 5 above)
Comment (alex-golts) [reply fusedrug_examples/interaction/drug_target/affinity_prediction/PLM_DTI/test.py:47]: it sets `torch.backends.cudnn.benchmark` 
if True I think it allows cudnn to search for an optimized algorithm given your hardware
